File: 개인공부/새 스터디/11주차/Rank.py
#https://programmers.co.kr/learn/courses/30/lessons/72412

from bisect import bisect_left
from itertools import product

def solution(info, query):
    answer = [] #정답 저장
    language = {'cpp' : [0], 'java' : [1], 'python' : [2], '-' : [0, 1, 2]}
    part = {'backend' : [0], 'frontend' : [1], '-' : [0, 1]}
    career = {'junior' : [0], 'senior' : [1], '-' : [0, 1]}
    food = {'chicken' : [0], 'pizza' : [1], '-' : [0, 1]}
    #그룹별로 저장, ex) group[0][0][0][0] = cpp쓰고 backend이고 junior이고 chicken 좋아함
    group = [[[[[] for _ in range(len(food)-1)] for _ in range(len(career)-1)] for _ in range(len(part)-1)] for _ in range(len(language)-1)]

    for _info in info:
        data = _info.split()
        group[language[data[0]][0]][part[data[1]][0]][career[data[2]][0]][food[data[3]][0]].append(int(data[4]))

    #그룹별로 저장한 리스트 정렬
    for p in product(range(len(language)-1), range(len(part)-1), range(len(career)-1), range(len(food)-1)):
        group[p[0]][p[1]][p[2]][p[3]].sort()

    #쿼리 전처리
    for q in query:
        cnt = 0
        raw_query = q.split()
        r_l, r_p, r_c, r_f = raw_query[::2]    #언어, 직군, 경력, 음식, 점수
        r_s = int(raw_query[-1])

        #카티션 프로덕트
        for p in product(language[r_l], part[r_p], career[r_c], food[r_f]):
            # 그룹별 리스트의 길이 - 요구하는 점수보다 높은 점수가 제일 처음 나오는 위치 = 요구하는 점수보다 높은 지원자들의 개수
            cnt += len(group[p[0]][p[1]][p[2]][p[3]]) - bisect_left(group[p[0]][p[1]][p[2]][p[3]], r_s)

        answer.append(cnt)

    return answer

if __name__ == '__main__':
    solution(["java backend junior pizza 150",
              "python frontend senior chicken 210",
              "python frontend senior chicken 150",
              "cpp backend senior pizza 260",
              "java backend junior chicken 80",
              "python backend senior chicken 50"],
             ["java and backend and junior and pizza 100",
              "python and frontend and senior and chicken 200",
              "cpp and - and senior and pizza 250",
              "- and backend and senior and - 150",
              "- and - and - and chicken 100",
              "- and - and - and - 150"])

File: 알고리즘 스터디/개인공부/PrefixSum/Rank.java
/* https://school.programmers.co.kr/learn/courses/30/lessons/72412 */

import java.util.*;

public class Rank {
    public int[] solution(String[] info, String[] query) {
        List<Integer> answer = new ArrayList<>();
        int[][][][][] applicant = new int[4][3][3][3][100001];
        HashMap<String, int[]> lanMap = new HashMap<>(4);
        HashMap<String, int[]> posMap = new HashMap<>(3);
        HashMap<String, int[]> carMap = new HashMap<>(3);
        HashMap<String, int[]> fooMap = new HashMap<>(3);

        lanMap.put("cpp", new int[]{0});
        lanMap.put("java", new int[]{1});
        lanMap.put("python", new int[]{2});
        lanMap.put("-", new int[]{0, 1, 2});

        posMap.put("backend", new int[]{0});
        posMap.put("frontend", new int[]{1});
        posMap.put("-", new int[]{0, 1});

        carMap.put("junior", new int[]{0});
        carMap.put("senior", new int[]{1});
        carMap.put("-", new int[]{0, 1});

        fooMap.put("chicken", new int[]{0});
        fooMap.put("pizza", new int[]{1});
        fooMap.put("-", new int[]{0, 1});

        for (int i1 = 0; i1 < applicant.length; i1++)
            for (int i2 = 0; i2 < applicant[i1].length; i2++)
                for (int i3 = 0; i3 < applicant[i1][i2].length; i3++)
                    for (int i4 = 0; i4 < applicant[i1][i2][i3].length; i4++)
                        Arrays.fill(applicant[i1][i2][i3][i4], 0);

        for (String str : info) {
            String[] tmp = str.split(" ");

            for (int i1 : lanMap.get(tmp[0]))
                for (int i2 : posMap.get(tmp[1]))
                    for (int i3 : carMap.get(tmp[2]))
                        for (int i4 : fooMap.get(tmp[3]))
                            applicant[i1][i2][i3][i4][Integer.parseInt(tmp[4])]++;

        }

        for (int i1 = 0; i1 < applicant.length; i1++)
            for (int i2 = 0; i2 < applicant[i1].length; i2++)
                for (int i3 = 0; i3 < applicant[i1][i2].length; i3++)
                    for (int i4 = 0; i4 < applicant[i1][i2][i3].length; i4++)
                        for (int i5 = applicant[i1][i2][i3][i4].length - 2; i5 >= 0; i5--)
                            applicant[i1][i2][i3][i4][i5] += applicant[i1][i2][i3][i4][i5 + 1];

        for (String str : query) {
            String[] tmp = str.split(" ");
            int result = 0;

            for (int i1 : lanMap.get(tmp[0]))
                for (int i2 : posMap.get(tmp[2]))
                    for (int i3 : carMap.get(tmp[4]))
                        for (int i4 : fooMap.get(tmp[6]))
                            result += applicant[i1][i2][i3][i4][Integer.parseInt(tmp[7])];

            answer.add(result);
        }

        return answer.stream().mapToInt(Integer::intValue).toArray();
    }

    public static void main(String[] args) {
        new Rank().solution(new String[]{"java backend junior pizza 150", "python frontend senior chicken 210", "python frontend senior chicken 150", "cpp backend senior pizza 260", "java backend junior chicken 80", "python backend senior chicken 50"},
                new String[]{"java and backend and junior and pizza 100", "python and frontend and senior and chicken 200", "cpp and - and senior and pizza 250", "- and backend and senior and - 150", "- and - and - and chicken 100", "- and - and - and - 150"});
        // [1,1,1,1,2,4]
    }
}
